test/extensions/tracers/xray/localized_sampling_test.cc

```diff
@@ -0,0 +1,532 @@
+      "url_path": "/api/move/*",
+      "fixed_target": 0
+    }
+  ],
+  "default": {
+    "fixed_target": 1,
+    "rate": 0.1
+  }
+}
+  )EOF";
+  LocalizedSamplingStrategy strategy{rules_json, random_generator, time_system_};
+  ASSERT_TRUE(strategy.usingDefaultManifest());
+}
+
+TEST_F(LocalizedSamplingStrategyTest, CustomRuleNegativeFixedRate) {
+  NiceMock<Runtime::MockRandomGenerator> random_generator;
+  constexpr auto rules_json = R"EOF(
+{
+  "version": 2,
+  "rules": [
+    {
+      "description": "X-Ray rule",
+      "host": "*",
+      "http_method": "*",
+      "url_path": "/api/move/*",
+      "fixed_target": -1,
+      "rate": 0.05
+    }
+  ],
+  "default": {
+    "fixed_target": 1,
+    "rate": 0.1
+  }
+}
+  )EOF";
+  LocalizedSamplingStrategy strategy{rules_json, random_generator, time_system_};
+  ASSERT_TRUE(strategy.usingDefaultManifest());
+}
+
+TEST_F(LocalizedSamplingStrategyTest, CustomRuleNegativeRate) {
+  NiceMock<Runtime::MockRandomGenerator> random_generator;
+  constexpr auto rules_json = R"EOF(
+{
+  "version": 2,
+  "rules": [
+    {
+      "description": "X-Ray rule",
+      "host": "*",
+      "http_method": "*",
+      "url_path": "/api/move/*",
+      "fixed_target": 0,
+      "rate": 0.05
+    }
+  ],
+  "default": {
+    "fixed_target": 1,
+    "rate": -0.1
+  }
+}
+  )EOF";
+  LocalizedSamplingStrategy strategy{rules_json, random_generator, time_system_};
+  ASSERT_TRUE(strategy.usingDefaultManifest());
+}
+
+TEST_F(LocalizedSamplingStrategyTest, TraceOnlyFromReservoir) {
+  NiceMock<Runtime::MockRandomGenerator> rng;
+  EXPECT_CALL(rng, random()).WillRepeatedly(Return(90));
+  constexpr auto rules_json = R"EOF(
+{
+  "version": 2,
+  "rules": [
+    {
+      "description": "X-Ray rule",
+      "host": "*",
+      "http_method": "*",
+      "url_path": "*",
+      "fixed_target": 1,
+      "rate": 0.5
+    }
+  ],
+  "default": {
+    "fixed_target": 1,
+    "rate": 0.5
+  }
+}
+  )EOF";
+
+  LocalizedSamplingStrategy strategy{rules_json, rng, time_system_};
+  ASSERT_FALSE(strategy.usingDefaultManifest());
+
+  SamplingRequest req;
+  ASSERT_TRUE(strategy.shouldTrace(req)); // first one should be traced
+  int i = 10;
+  while (i-- > 0) {
+    ASSERT_FALSE(strategy.shouldTrace(req));
+  }
+}
+
+TEST_F(LocalizedSamplingStrategyTest, TraceFromReservoirAndByRate) {
+  NiceMock<Runtime::MockRandomGenerator> rng;
+  EXPECT_CALL(rng, random()).WillRepeatedly(Return(1));
+  constexpr auto rules_json = R"EOF(
+{
+  "version": 2,
+  "rules": [
+    {
+      "description": "X-Ray rule",
+      "host": "*",
+      "http_method": "*",
+      "url_path": "*",
+      "fixed_target": 1,
+      "rate": 0.1
+    }
+  ],
+  "default": {
+    "fixed_target": 1,
+    "rate": 0.5
+  }
+}
+  )EOF";
+
+  LocalizedSamplingStrategy strategy{rules_json, rng, time_system_};
+  ASSERT_FALSE(strategy.usingDefaultManifest());
+
+  SamplingRequest req;
+  ASSERT_TRUE(strategy.shouldTrace(req)); // first one should be traced
```

Comment:
This comment probably needs to be updated since it appears that all requests are traced

Reply:
I'll remove the whole line. It's meaningless in this case.